recipe/session/cookieAndHeaders.go

```diff
@@ -316,3 +318,68 @@ func getCookieName(cookie string) string {
 	}
 	return kv[0]
 }
+
+// ClearSessionCookiesFromOlderCookieDomain addresses an edge case where changing the cookieDomain config on the server can
+// lead to session integrity issues. For instance, if the API server URL is 'api.example.com'
+// with a cookie domain of '.example.com', and the server updates the cookie domain to 'api.example.com',
+// the client may retain cookies with both '.example.com' and 'api.example.com' domains.
+//
+// Consequently, if the server chooses the older cookie, session invalidation occurs, potentially
+// resulting in an infinite refresh loop. To fix this, users are asked to specify "OlderCookieDomain" in
+// the config.
+//
+// This function checks for multiple cookies with the same name and clears the cookies for the older domain.
+func ClearSessionCookiesFromOlderCookieDomain(req *http.Request, res http.ResponseWriter, config sessmodels.TypeNormalisedInput, userContext supertokens.UserContext) error {
+	allowedTransferMethod := config.GetTokenTransferMethod(req, false, userContext)
+
+	// If the transfer method is 'header', there's no need to clear cookies immediately, even if there are multiple in the request.
+	if allowedTransferMethod == sessmodels.HeaderTransferMethod {
+		return nil
+	}
+
+	didClearCookies := false
+
+	tokenTypes := []sessmodels.TokenType{sessmodels.AccessToken, sessmodels.RefreshToken}
+	for _, token := range tokenTypes {
+		if hasMultipleCookiesForTokenType(req, token) {
+			// If a request has multiple session cookies and 'olderCookieDomain' is
+			// unset, we can't identify the correct cookie for refreshing the session.
+			// Using the wrong cookie can cause an infinite refresh loop. To avoid this,
+			// we throw a 500 error asking the user to set 'olderCookieDomain'.
+			if config.OlderCookieDomain == nil {
+				return errors.New(`The request contains multiple session cookies. This may happen if you've changed the 'cookieDomain' value in your configuration. To clear tokens from the previous domain, set 'olderCookieDomain' in your config.`)
+			}
+
+			supertokens.LogDebugMessage(fmt.Sprint("ClearSessionCookiesFromOlderCookieDomain: Clearing duplicate ", token, " cookie with domain ", config.OlderCookieDomain))
+			config.CookieDomain = config.OlderCookieDomain
```

Comment:
not a good idea to change the config. We should instead pass the cookieDoamin value to the setToken function similarly to whats done in python

Reply:
Unless you are sure that this is modifying only a copy of the config and not the actual config. Please double check and confirm this.

Reply:
Yeah, this was the first thing I checked when reviewing the PR. It only modifies the copy.